lib/krb5/get_cred.c

```diff
@@ -1193,6 +1193,8 @@ _krb5_get_cred_kdc_any(krb5_context context,
 {
     krb5_error_code ret;
     krb5_deltat offset;
+    krb5_boolean no_ref = in_creds->server->name.name_string.len < 2 &&
+			  !flags.b.canonicalize;
 
```

Comment:
`need_referral()` in the KDC allows referrals to be sent for one-part names. Can you describe the desired behavior in more detail? Will Windows issue a ticket when the SPN is specified as `sAMAccountName$@REALM`?

Reply:
This looks like an optimization: if the capaths path didn't work, why would the referrals path work if the KDC wouldn't issue a referral?  And right now Heimdal (and MIT) only do referrals in TGS when the service principal is something like a host-based service name.  But I don't think we should preclude referrals for other name types, so I'm inclined to reject this part of this change.

Reply:
Agreed

Reply:
Yes, that's right, wouldn't call it an optimization but to preserve the error (such client not found) if we know it will certainly fail.

The background as mentioned, is that the change of order between referral and capath functions broke samba torture suite miserably. I was working on adapting the tests (while keeping the actual check of the KDC behavior) and noticed that sometimes KDC error from capath get masked by the error in the referral fallback (matching cred not found) and thought we could improve that.

Reply:
Ah, sorry @nicowilliams I think I misread your comment previously. I agree that we might want to allow referrals for other name types but right now the referral function would error immediately if the name is short and canonicalize was not requested which hides an actual error from the KDC.
Perhaps I should move this check to a function which would be called from both places.

Reply:
Ohhh, I get it now.  The problem is that `get_cred_kdc_referral()` does this very same check and returns an error that is useless because it clobbers the more useful error from `get_cred_kdc_capath()`.